sdk/core/core-rest-pipeline/src/policies/multipartPolicy.ts

```diff
@@ -0,0 +1,153 @@
+import { toStream, concatenateStreams } from "../util/stream";
+import { isBlob } from "../util/typeGuards";
+
+function generateBoundary(): string {
+  return `----AzSDKFormBoundary${randomUUID()}`;
+}
+
+function encodeHeaders(headers: HttpHeaders): string {
+  let result = "";
+  for (const [key, value] of headers) {
+    result += `${key}: ${value}\r\n`;
+  }
+  return result;
+}
+
+function getLength(
+  source:
+    | (() => ReadableStream<Uint8Array>)
+    | (() => NodeJS.ReadableStream)
+    | Uint8Array
+    | Blob
+    | ReadableStream
+    | NodeJS.ReadableStream
+): number | undefined {
+  if (source instanceof Uint8Array) {
+    return source.byteLength;
+  } else if (isBlob(source)) {
+    // if was created using createFile then -1 means we have an unknown size
+    return source.size === -1 ? undefined : source.size;
+  } else {
+    return undefined;
+  }
+}
+
+function getTotalLength(
+  sources: (
+    | (() => ReadableStream<Uint8Array>)
+    | (() => NodeJS.ReadableStream)
+    | Uint8Array
+    | Blob
+    | ReadableStream
+    | NodeJS.ReadableStream
+  )[]
+): number | undefined {
+  let total = 0;
+  for (const source of sources) {
+    const partLength = getLength(source);
+    if (partLength === undefined) {
+      return undefined;
+    } else {
+      total += partLength;
+    }
+  }
+  return total;
+}
+
+function buildRequestBody(request: PipelineRequest, parts: BodyPart[], boundary: string): void {
+  const sources = [
+    stringToUint8Array(`--${boundary}`, "utf-8"),
+    ...parts.flatMap((part) => [
+      stringToUint8Array("\r\n", "utf-8"),
+      stringToUint8Array(encodeHeaders(part.headers), "utf-8"),
+      stringToUint8Array("\r\n", "utf-8"),
+      part.body,
+      stringToUint8Array(`\r\n--${boundary}`, "utf-8"),
+    ]),
+    stringToUint8Array("--\r\n\r\n", "utf-8"),
+  ];
+
+  const contentLength = getTotalLength(sources);
+  if (contentLength) {
+    request.headers.set("Content-Length", contentLength);
+  }
+
+  request.body = (() =>
+    concatenateStreams(
+      sources.map((source) => (typeof source === "function" ? source() : source)).map(toStream)
+    )) as RequestBodyType;
+}
+
+/**
+ * Name of multipart policy
+ */
+export const multipartPolicyName = "multipartPolicy";
+
+const maxBoundaryLength = 70;
+const validBoundaryCharacters = new Set(
+  `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'()+,-./:=?`
+);
+
+function assertValidBoundary(boundary: string): void {
+  if (boundary.length > maxBoundaryLength) {
+    throw new Error(`Multipart boundary "${boundary}" exceeds maximum length of 70 characters`);
+  }
+
+  if (Array.from(boundary).some((x) => !validBoundaryCharacters.has(x))) {
+    throw new Error(`Multipart boundary "${boundary}" contains invalid characters`);
+  }
+}
+
+/**
+ * Pipeline policy for multipart requests
+ */
+export function multipartPolicy(): PipelinePolicy {
+  return {
+    name: multipartPolicyName,
```

Comment:
Just curious why we split the form-data policy and multipart policy?

Reply:
There are other kinds of multipart content, e.g. `multipart/alternative` and `multipart/mixed`. Separating the two allows for us to support those content types in future -- we already have instances of SDKs constructing the multipart body themselves in these situations, and in future they won't have to do that.